### PR TITLE
feat: introduce ui tests based on vscode-extension-tester

### PR DESCRIPTION
**Pull request by odockal**
Signed-off-by: Ondrej Dockal <odockal@redhat.com>

Partial-Fix: #139

### CODE REVIEW

**Review by ssbarnea — APPROVED** [review]
I also manually downloaded the pr and run the test locally and they worked as expected with `npm run ui-test`.

**Comment by odockal** [comment]
@ssbarnea I had to update npm script "webpack" with `npm run compile` in order to compile the ts classes on out folder. also their placement in out/client/ui-test is no quite fortunate. It all could be probably done better.

**Comment by ssbarnea** [comment]
> @ssbarnea I had to update npm script "webpack" with `npm run compile` in order to compile the ts classes in out folder. also their placement in out/client/ui-test is no quite fortunate. It all could be probably done better.

While trying to setup chai, I tried to use the `test/` root repository but I did not know all the [tricks](https://stackoverflow.com/questions/59042716/trying-es6-style-import-gives-cannot-use-import-statement-outside-a-module) required for making ts compile automatically. If we could add a follow-up to enable us to use `npx mocha` directly it would be even better.

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -49,6 +49,12 @@ jobs:
         with:
           run: npm test
 
+      # Run UI tests
+      - name: Run test
+        uses: GabrielBB/xvfb-action@v1.0
+        with:
+          run: npm run ui-test
```
Comment (webknjaz) [inline .github/workflows/ci.yaml:56]: This should go into a separate factor in the matrix. At least until the team agrees on how to handle it.

**Review by webknjaz — CHANGES_REQUESTED** [review]
I like it, except for one bit below.

File: src/ui-test/extensionUITest.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        const actionButton = await titlePart.getAction(
+          'Clear Extensions Search Results'
+        );
+        if (await actionButton.isEnabled()) {
+          await actionButton.click();
+        }
+      }
+      if (quickBox && (await quickBox.isDisplayed())) {
+        await quickBox.cancel();
+      }
```
Comment (webknjaz) [inline src/ui-test/extensionUITest.ts:54]:
```suggestion
      if (!quickBox) {
        return;
      }

      await quickBox.isDisplayed();
      await quickBox.cancel();
```

File: src/ui-test/extensionUITest.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+} from 'vscode-extension-tester';
+
+/**
+ * @author Ondrej Dockal <odockal@redhat.com>
+ */
+export function extensionUIAssetsTest(): void {
+  describe('Verify base assets are available after installation', () => {
+    let view: ViewControl;
+    let sideBar: SideBarView;
+    let quickBox: InputBox;
```
Comment (webknjaz) [inline src/ui-test/extensionUITest.ts:17]: It's not clear why this exists. It seems like in every code path it's `undefined`.

File: src/ui-test/extensionUITest.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    before(async function () {
```
Comment (webknjaz) [inline src/ui-test/extensionUITest.ts:19]: Nit: why isn't it an arrow function? This is inconsistent with the outer lambda. Does `this` need to be bound to a different context?

Comment (odockal) [reply src/ui-test/extensionUITest.ts:19]: It is mocha thing, arrow function wont let you get into mocha context to set timeout for the test: https://mochajs.org/#arrow-functions

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (odockal) [reply .github/workflows/ci.yaml]: I have removed it, please go ahead and set it on your own the way you prefer.

File: src/ui-test/extensionUITest.ts (hunk 3 above)
Comment (odockal) [reply src/ui-test/extensionUITest.ts]: very good point, unused it is indeed. Removing.

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -17,7 +17,11 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
+        npm-target:
+        - test
+        - ui-test
         os: [macos-latest, windows-latest, ubuntu-latest]
+
```
Comment (webknjaz) [inline .github/workflows/ci.yaml:24]:
```suggestion
```

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (webknjaz) [reply .github/workflows/ci.yaml]: done

**Comment by ssbarnea** [comment]
@webknjaz Shortly `npm test` is just a [shortcut to `npm run test`](https://stackoverflow.com/questions/57031559/difference-between-npm-t-and-npm-run-test#:~:text=TL%3BDR%20there%20is%20no,same%20action%20in%20this%20case.) and we had no real test running.

#204 added the first part of extest preparation of vscode and installation of the extension, this channels is the follow-up for completing the test enablement. The are no reasons to have different commands, especially as you cannot really test the extension without installing it with extest.

File: package.json [hunk 6]
```diff
@@ -407,10 +408,11 @@
     "prepare": "husky install",
     "pretest": "npm run compile && npm run lint",
     "test": "extest get-vscode && extest install-vsix",
+    "ui-test": "extest setup-and-run -i out/client/ui-test/allTestsSuite.js",
     "vscode:prepublish": "npm run webpack",
     "watch": "tsc -b -w",
     "watch:server": "tsc -p ../ansible-language-server --outDir out/server -w",
-    "webpack": "npm run clean && webpack --mode production --config ./webpack.config.js",
+    "webpack": "npm run clean && npm run compile && webpack --mode production --config ./webpack.config.js",
```
Comment (tomaciazek) [inline package.json:415]: Why do we need to compile first? Webpack does compilation too.

**Review by tomaciazek — APPROVED** [review]
LGTM. Just one comment, but not a showstopper.

File: package.json (hunk 6 above)
Comment (odockal) [reply package.json:415]: Webpack does not compile ui test related files (*.ts) into *.js. It places only *.ts file under out folder.

Comment (odockal) [reply package.json:415]: So webpack-like compilation differs from what ```npm run compile``` does.

**Comment by webknjaz** [comment]
> @webknjaz Shortly `npm test` is just a [shortcut to `npm run test`](https://stackoverflow.com/questions/57031559/difference-between-npm-t-and-npm-run-test#:~:text=TL%3BDR%20there%20is%20no,same%20action%20in%20this%20case.) and we had no real test running.

Oh, I know about this alias. I think we could have lightweight unit tests there and have heavy UI tests separately. This could be debated separately.

File: package.json (hunk 6 above)
Comment (webknjaz) [reply package.json:415]: I'm going to mark this as resolved to unblock the release.

**Review by webknjaz — APPROVED** [review]

**Pull request merged**